### PR TITLE
Fix the range of categorical values.

### PR DESCRIPTION
**Pull request by toshihikoyanase**
## Motivation
This PR addresses https://github.com/optuna/optuna/issues/1967.

## Description of the changes
This PR changes the plot range of categorical parameters for `plotly>=4.12.0`. As I mentioned in the comment, plotly seems to generate contours using indices of categorical values, but it determines plot ranges based on raw values. So, I updated the range based on the indices.

Code (from #1819)
```python
import optuna

def objective(trial):
    x = trial.suggest_categorical("x", ["10", "11", "100"])
    y = trial.suggest_categorical("y", ["sigmoid", "relu", "leakyrelu"])
    z = trial.suggest_loguniform("z", 1e-1, 1e2)
    return int("0" in x) + abs(len(y) - 5) + (z/50) ** 2

study = optuna.create_study()
study.optimize(objective, n_trials=30)

optuna.visualization.plot_contour(study).show()
```

This PR:
image

The current master:
image

### CODE REVIEW

**Review by hvy — COMMENTED** [review]
Thanks, basically LGTM. I checked locally with `4.12.0` and `<4.12.0`, with and without these  changes. Just wondering if it is difficult to keep the contours all the way to the edges as previously, [docs](https://optuna.readthedocs.io/en/v2.3.0/reference/visualization/generated/optuna.visualization.plot_contour.html)?

**Comment by toshihikoyanase** [comment]
@hvy Sorry for the late reply.
I'm not sure why the new version does not render the edges. I personally feel we can work on it in a follow-up PR to fix the current document soon.

CC @harupy @ytknzw Please let me know if you have any comments on this issue.

File: optuna/visualization/_contour.py [hunk 1]
```diff
@@ -109,15 +111,23 @@ def _get_contour_plot(study: Study, params: Optional[List[str]] = None) -> "go.F
...
             padding = (math.log10(max_value) - math.log10(min_value)) * padding_ratio
             min_value = math.pow(10, math.log10(min_value) - padding)
             max_value = math.pow(10, math.log10(max_value) + padding)
 
-        elif _is_categorical(trials, p_name):
-            # For numeric values, plotly does not automatically plot as "category" type.
-            update_category_axes[p_name] = any([str(v).isnumeric() for v in set(values)])
-
         else:
             padding = (max_value - min_value) * padding_ratio
```
Comment (hvy) [inline optuna/visualization/_contour.py:132]: This code is executed now with categoricals without numerical conversions for `plotly<=4.11.0`. I think it'll lead to an error.

Comment (toshihikoyanase) [reply optuna/visualization/_contour.py:132]: Thank you for pointing it out. I'll check it.

Comment (toshihikoyanase) [reply optuna/visualization/_contour.py:134]: As you mentioned it raised the error when I used `plotly==4.11.0`. I fixed the error in the commit https://github.com/optuna/optuna/pull/1983/commits/5ad0bbb803ab8557887bb0099eacf9b4a02c67f9.

**Comment by toshihikoyanase** [comment]
@hvy @himkt I resolved the issue with `plotly<4.12.0`. PTAL!

**Review by hvy — APPROVED** [review]
Thanks for the fix. Quickly verified this with different plotly versions locally. LGTM.

It wasn't entire clear to me what the error is referring to in the TODO comment, though.

**Comment by toshihikoyanase** [comment]
@hvy I read the comment again, and I don't think it's a TODO comment. So, I revised it as follows in https://github.com/optuna/optuna/pull/1983/commits/ddf7cfd9f2cd8a588ef3e818acdd62feef6aa2ff:

> Plotly>=4.12.0 draws contours using the indices of categorical variables instead of
> raw values and the range should be updated based on the cardinality of categorical
> variables. See https://github.com/optuna/optuna/issues/1967.

**Review by himkt — APPROVED** [review]
Thank you for the update, LGTM.

> I'm not sure why the new version does not render the edges.

Let's modify it when we find the cause.

**Pull request merged**